File: go/internal/codex/codex.go
package codex

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"

	"codex-web-mcp/internal/config"
)

func Run(ctx context.Context, prompt string) string {
	tmp, err := os.CreateTemp("", "codex-out-*.txt")
	if err != nil {
		return fmt.Sprintf("ERROR: failed to create temp file: %v", err)
	}
	tmpPath := tmp.Name()
	_ = tmp.Close()
	defer os.Remove(tmpPath)

	timeoutCtx, cancel := context.WithTimeout(ctx, time.Duration(config.CodexTimeoutSeconds)*time.Second)
	defer cancel()

	args := []string{
		"exec",
		"--skip-git-repo-check",
		"--ephemeral",
		"--color", "never",
		"--sandbox", "read-only",
		"--output-last-message", tmpPath,
		prompt,
	}
	cmd := exec.CommandContext(timeoutCtx, config.CodexBin(), args...)
	cmd.Stdin = strings.NewReader("")
	cmd.Stdout = nil // discard
	var stderr strings.Builder
	cmd.Stderr = &stderrCapture{w: &stderr}

	if err := cmd.Run(); err != nil {
		if timeoutCtx.Err() == context.DeadlineExceeded {
			return "ERROR: codex timed out after 180s"
		}
		return fmt.Sprintf("ERROR: codex exited with error: %v: %s", err, strings.TrimSpace(stderr.String()))
	}

	info, err := os.Stat(tmpPath)
	if err != nil {
		return "ERROR: codex produced no output file"
	}
	if info.Size() == 0 {
		return "ERROR: codex output file empty"
	}
	data, err := os.ReadFile(tmpPath)
	if err != nil {
		return fmt.Sprintf("ERROR: failed to read codex output: %v", err)
	}
	s := strings.TrimRight(string(data), " \t\r\n")
	if s == "" {
		return "ERROR: codex output file empty"
	}
	return s
}

type stderrCapture struct {
	w *strings.Builder
}

func (s *stderrCapture) Write(p []byte) (int, error) {
	s.w.Write(p)
	return len(p), nil
}

File: fsharp/src/CodexWebMcp.FSharp/Codex.fs
module CodexWebMcp.FSharp.Codex

open System
open System.Diagnostics
open System.IO
open System.Text
open System.Threading
open System.Threading.Tasks

let runAsync (prompt: string) : Task<string> =
    task {
        let tmpPath =
            Path.Combine(Path.GetTempPath(), sprintf "codex-out-%s.txt" (Guid.NewGuid().ToString("N")))

        let psi = ProcessStartInfo()
        psi.FileName <- Config.codexBin ()
        psi.RedirectStandardInput <- true
        psi.RedirectStandardOutput <- true
        psi.RedirectStandardError <- true
        psi.UseShellExecute <- false
        psi.CreateNoWindow <- true
        for a in [
            "exec"
            "--skip-git-repo-check"
            "--ephemeral"
            "--color"; "never"
            "--sandbox"; "read-only"
            "--output-last-message"; tmpPath
            prompt
        ] do psi.ArgumentList.Add(a)

        let proc = new Process()
        proc.StartInfo <- psi

        try
            try
                if not (proc.Start()) then
                    return "ERROR: codex invocation failed: Start returned false"
                else
                    proc.StandardInput.Close()

                    let stdoutTask = proc.StandardOutput.ReadToEndAsync()
                    let stderrTask = proc.StandardError.ReadToEndAsync()

                    use cts = new CancellationTokenSource(TimeSpan.FromSeconds(float Config.CodexTimeoutSeconds))
                    let! exited =
                        task {
                            try
                                do! proc.WaitForExitAsync(cts.Token)
                                return true
                            with :? OperationCanceledException ->
                                return false
                        }

                    if not exited then
                        try proc.Kill(entireProcessTree = true) with _ -> ()
                        return "ERROR: codex timed out after 180s"
                    else
                        let! _ = stdoutTask
                        let! stderr = stderrTask
                        if proc.ExitCode <> 0 then
                            return sprintf "ERROR: codex exited with code %d: %s" proc.ExitCode (stderr.Trim())
                        elif not (File.Exists tmpPath) then
                            return "ERROR: codex produced no output file"
                        else
                            let fi = FileInfo(tmpPath)
                            if fi.Length = 0L then
                                return "ERROR: codex output file empty"
                            else
                                let data = (File.ReadAllText(tmpPath)).TrimEnd([| ' '; '\t'; '\r'; '\n' |])
                                if data = "" then return "ERROR: codex output file empty"
                                else return data
            with ex ->
                return sprintf "ERROR: codex invocation failed: %s" ex.Message
        finally
            try File.Delete(tmpPath) with _ -> ()
            proc.Dispose()
    }
